File: soul-stream/src/nodes/node-connection.ts
/**
 * NodeConnection — 개별 소울 서버 노드의 WebSocket 연결 관리.
 *
 * 노드에 명령을 전송하고, 노드에서 오는 이벤트를 수신하여 구독자에게 전달한다.
 */

import { randomUUID } from "crypto";
import type { WebSocket } from "ws";
import type { NodeRegistration, NodeInfo, NodeStatus } from "./types";
import type { SessionSummary, SessionEvent } from "../sessions/types";

let requestIdCounter = 0;
function nextRequestId(): string {
  return `req-${++requestIdCounter}-${Date.now()}`;
}

export class NodeConnection {
  readonly nodeId: string;
  readonly host: string;
  readonly port: number;
  readonly capabilities: Record<string, unknown>;
  readonly connectedAt: number;

  private _ws: WebSocket;
  private _status: NodeStatus = "connected";
  private _sessions: Map<string, SessionSummary> = new Map();
  private _eventListeners: Map<string, Set<(event: SessionEvent) => void>> =
    new Map();
  /** subscribeEvents 리스너: (event, eventId) 콜백 */
  private _subscribeListeners: Map<
    string,
    Map<string, (event: SessionEvent, eventId: number) => void>
  > = new Map();
  /** subscribeEvents 요청 추적: subscribeId → sessionId (에러 응답 라우팅용) */
  private _subscribeRequestMap: Map<string, string> = new Map();
  private _pendingRequests: Map<
    string,
    { resolve: (data: unknown) => void; reject: (err: Error) => void }
  > = new Map();
  private _onClose: (() => void) | null = null;

  constructor(ws: WebSocket, registration: NodeRegistration) {
    this.nodeId = registration.node_id;
    this.host = registration.host;
    this.port = registration.port;
    this.capabilities = registration.capabilities ?? {};
    this.connectedAt = Date.now();
    this._ws = ws;

    this._setupListeners();
  }

  get status(): NodeStatus {
    return this._status;
  }

  /** 연결 종료 콜백 등록 (NodeManager가 사용). */
  set onClose(cb: (() => void) | null) {
    this._onClose = cb;
  }

  /** 노드 정보 반환. */
  toInfo(): NodeInfo {
    return {
      nodeId: this.nodeId,
      host: this.host,
      port: this.port,
      status: this._status,
      capabilities: this.capabilities,
      connectedAt: this.connectedAt,
      sessionCount: this._sessions.size,
    };
  }

  /** 세션 목록 반환. */
  getSessions(): SessionSummary[] {
    return Array.from(this._sessions.values());
  }

  /** soul-server HTTP 베이스 URL 반환. 0.0.0.0은 127.0.0.1로 정규화. */
  getHttpBaseUrl(): string {
    const host = this.host === "0.0.0.0" ? "127.0.0.1" : this.host;
    return `http://${host}:${this.port}`;
  }

  /** 세션 생성 명령 전송. */
  async createSession(
    prompt: string,
    opts?: {
      profile?: string;
      allowed_tools?: string[];
      disallowed_tools?: string[];
      use_mcp?: boolean;
    }
  ): Promise<string> {
    const requestId = nextRequestId();
    const result = await this._sendCommand({
      type: "create_session",
      prompt,
      profile: opts?.profile ?? "",
      allowed_tools: opts?.allowed_tools,
      disallowed_tools: opts?.disallowed_tools,
      use_mcp: opts?.use_mcp,
      request_id: requestId,
    });
    return (result as { session_id: string }).session_id;
  }

  /** 개입 명령 전송. */
  async intervene(
    sessionId: string,
    text: string,
    user: string
  ): Promise<void> {
    this._send({
      type: "intervene",
      session_id: sessionId,
      text,
      user,
    });
  }

  /** AskUserQuestion 응답 전송. */
  async respond(
    sessionId: string,
    requestId: string,
    answers: Record<string, unknown>
  ): Promise<void> {
    this._send({
      type: "respond",
      session_id: sessionId,
      request_id: requestId,
      answers,
    });
  }

  /** 세션 이벤트 구독 (기존 onSessionEvent). 해제 함수를 반환한다. */
  onSessionEvent(
    sessionId: string,
    listener: (event: SessionEvent) => void
  ): () => void {
    let listeners = this._eventListeners.get(sessionId);
    if (!listeners) {
      listeners = new Set();
      this._eventListeners.set(sessionId, listeners);
    }
    listeners.add(listener);

    return () => {
      listeners!.delete(listener);
      if (listeners!.size === 0) {
        this._eventListeners.delete(sessionId);
      }
    };
  }

  /**
   * subscribe_events 커맨드를 soul-server에 전송하고 라이브 이벤트를 수신한다.
   *
   * @param sessionId 구독할 세션 ID
   * @param afterId DB 재생 시작 커서 (0이면 처음부터)
   * @param listener (event, eventId) 콜백
   * @returns 구독 해제 함수
   */
  subscribeEvents(
    sessionId: string,
    afterId: number,
    listener: (event: SessionEvent, eventId: number) => void
  ): () => void {
    const subscribeId = randomUUID();

    // 세션별 리스너 맵에 등록
    let sessionListeners = this._subscribeListeners.get(sessionId);
    if (!sessionListeners) {
      sessionListeners = new Map();
      this._subscribeListeners.set(sessionId, sessionListeners);
    }
    sessionListeners.set(subscribeId, listener);
    this._subscribeRequestMap.set(subscribeId, sessionId);

    // soul-server에 subscribe_events 커맨드 전송
    this._send({
      type: "subscribe_events",
      session_id: sessionId,
      after_id: afterId,
      request_id: subscribeId,
    });

    return () => {
      const map = this._subscribeListeners.get(sessionId);
      if (map) {
        map.delete(subscribeId);
        if (map.size === 0) {
          this._subscribeListeners.delete(sessionId);
        }
      }
      this._subscribeRequestMap.delete(subscribeId);
    };
  }

  /** WebSocket 연결 닫기. */
  close(): void {
    this._status = "disconnected";
    try {
      this._ws.close();
    } catch {
      // 이미 닫혀있을 수 있음
    }
  }

  // ─── Private ──────────────────────────────────

  private _setupListeners(): void {
    this._ws.on("message", (data) => {
      try {
        const msg = JSON.parse(data.toString());
        this._handleMessage(msg);
      } catch {
        // invalid JSON — ignore
      }
    });

    this._ws.on("close", () => {
      this._status = "disconnected";
      // pending requests 거부
      for (const [, pending] of this._pendingRequests) {
        pending.reject(new Error("Node disconnected"));
      }
      this._pendingRequests.clear();
      this._onClose?.();
    });

    this._ws.on("error", () => {
      // error 이벤트는 close 이벤트 전에 발생 — close에서 처리
    });
  }

  private _handleMessage(msg: Record<string, unknown>): void {
    const type = msg.type as string;
    const requestId = msg.request_id as string | undefined;

    // pending request 응답 처리
    if (requestId && this._pendingRequests.has(requestId)) {
      const pending = this._pendingRequests.get(requestId)!;
      this._pendingRequests.delete(requestId);

      if (type === "error") {
        pending.reject(new Error((msg.message as string) ?? "Unknown error"));
        return;
      }

      // session_created 응답이면 세션 목록에도 추가
      if (type === "session_created") {
        const sid = msg.session_id as string;
        if (sid) {
          this._sessions.set(sid, { sessionId: sid, status: "running" });
        }
      }

      pending.resolve(msg);
      return;
    }

    // subscribeEvents error 응답 처리
    if (type === "error" && requestId && this._subscribeRequestMap.has(requestId)) {
      // 에러 응답은 subscribeId로 식별. Map 순회 시 spread로 안전하게 처리.
      for (const [subId, sessionId] of [...this._subscribeRequestMap]) {
        if (subId === requestId) {
          this._subscribeListeners.get(sessionId)?.delete(subId);
          this._subscribeRequestMap.delete(subId);
          break;
        }
      }
      return;
    }

    switch (type) {
      case "event": {
        const sessionId = msg.session_id as string;
        const sessionEvent: SessionEvent = {
          type: msg.type as string,
          session_id: msg.session_id as string,
          event: msg.event as Record<string, unknown> | undefined,
          id: msg.event_id as number | undefined,
        };

        // 기존 onSessionEvent 리스너 (_eventListeners) — 제거하지 않음
        const listeners = this._eventListeners.get(sessionId);
        if (listeners) {
          for (const listener of listeners) {
            try {
              listener(sessionEvent);
            } catch {
              // 리스너 에러 무시
            }
          }
        }

        // subscribeEvents 리스너 (_subscribeListeners)
        this._subscribeListeners.get(sessionId)?.forEach(
          (listener) => listener(sessionEvent, sessionEvent.id ?? 0)
        );
        break;
      }

      case "sessions_update": {
        // 세션 목록 업데이트
        const sessions = msg.sessions as SessionSummary[];
        if (Array.isArray(sessions)) {
          this._sessions.clear();
          for (const s of sessions) {
            const id =
              s.sessionId ??
              (s as unknown as Record<string, unknown>).agent_session_id ??
              "";
            if (id) {
              // sessionId를 정규화하여 저장 (upstream이 agent_session_id만 보내는 경우 대비)
              this._sessions.set(id as string, { ...s, sessionId: id as string });
            }
          }
        }
        break;
      }

      case "session_created": {
        // 새 세션 추가
        const sessionId = msg.session_id as string;
        if (sessionId) {
          const sessionData = (msg.session as Record<string, unknown>) ?? {};
          this._sessions.set(sessionId, {
            ...(sessionData as unknown as Partial<SessionSummary>),
            sessionId,
            status: (sessionData.status as string) ?? "running",
          });
        }
        break;
      }

      case "session_updated": {
        // 세션 상태/메시지 갱신
        const id =
          (msg.agent_session_id as string) ??
          (msg.sessionId as string) ??
          "";
        if (id && this._sessions.has(id)) {
          const existing = this._sessions.get(id)!;
          this._sessions.set(id, {
            ...existing,
            ...(msg.status != null ? { status: msg.status as string } : {}),
            ...(msg.updated_at != null
              ? { updatedAt: msg.updated_at as string }
              : {}),
            ...(msg.last_message != null
              ? { lastMessage: msg.last_message as string }
              : {}),
          });
        }
        break;
      }

      case "session_deleted": {
        // 세션 제거
        const delId =
          (msg.agent_session_id as string) ??
          (msg.sessionId as string) ??
          "";
        if (delId) {
          this._sessions.delete(delId);
        }
        break;
      }

      case "health_status": {
        // 헬스 상태 — 현재는 로깅만
        break;
      }
    }
  }

  private _send(data: Record<string, unknown>): void {
    if (this._ws.readyState === this._ws.OPEN) {
      this._ws.send(JSON.stringify(data));
    }
  }

  private _sendCommand(
    data: Record<string, unknown>,
    timeoutMs = 30000
  ): Promise<unknown> {
    const requestId = data.request_id as string;

    return new Promise((resolve, reject) => {
      this._pendingRequests.set(requestId, { resolve, reject });
      this._send(data);

      // 타임아웃
      setTimeout(() => {
        if (this._pendingRequests.has(requestId)) {
          this._pendingRequests.delete(requestId);
          reject(new Error(`Request ${requestId} timed out`));
        }
      }, timeoutMs);
    });
  }
}

File: soulstream-server/src/soulstream_server/nodes/node_connection.py
"""
NodeConnection — soul-server 노드의 WebSocket 연결을 래핑.

노드 정보 추적, 명령 전송, 수신 메시지 디스패치, 응답 대기(Future) 관리.
"""

import asyncio
import json
import logging
import time
import uuid
from datetime import datetime, timezone
from typing import Any, Callable, Coroutine, Optional

from fastapi import WebSocket
from starlette.websockets import WebSocketDisconnect

from soulstream_server.constants import (
    CMD_CREATE_SESSION,
    CMD_INTERVENE,
    CMD_RESPOND,
    CMD_SUBSCRIBE_EVENTS,
    COMMAND_TIMEOUT,
    EVT_ERROR,
    EVT_EVENT,
    EVT_HEALTH_STATUS,
    EVT_SESSION_CREATED,
    EVT_SESSION_DELETED,
    EVT_SESSION_UPDATED,
    EVT_SESSIONS_UPDATE,
)

logger = logging.getLogger(__name__)

# Callback type aliases
OnCloseCallback = Callable[["NodeConnection"], Coroutine[Any, Any, None]]
OnSessionChangeCallback = Callable[
    [str, str, dict | None], Coroutine[Any, Any, None]
]  # (node_id, change_type, data)


class NodeConnection:
    """soul-server 노드 하나의 WebSocket 연결."""

    def __init__(
        self,
        ws: WebSocket,
        node_id: str,
        host: str = "",
        port: int = 0,
        capabilities: list[str] | None = None,
        on_close: OnCloseCallback | None = None,
        on_session_change: OnSessionChangeCallback | None = None,
    ):
        self._ws = ws
        self.node_id = node_id
        self.host = host
        self.port = port
        self.capabilities = capabilities or []
        self.connected_at = datetime.now(timezone.utc)

        self._sessions: dict[str, dict] = {}
        self._agent_profiles: dict = {}  # 연결 직후 _fetch_agent_profiles()로 populate됨
        self._portrait_cache: dict[str, bytes] = {}  # agent_id → portrait bytes (등록 메시지에서 수신)
        self._request_counter = 0
        self._pending: dict[str, asyncio.Future] = {}
        self._subscribe_listeners: dict[str, dict[str, Callable]] = {}

        self.on_close = on_close
        self.on_session_change = on_session_change

    @property
    def sessions(self) -> dict[str, dict]:
        return self._sessions

    @property
    def agent_profiles(self) -> dict:
        return self._agent_profiles

    @property
    def portrait_cache(self) -> dict[str, bytes]:
        return self._portrait_cache

    def set_agent_data(
        self, profiles: dict, portrait_cache: dict[str, bytes]
    ) -> None:
        """에이전트 프로필과 portrait 캐시를 설정한다."""
        self._agent_profiles = profiles
        self._portrait_cache = portrait_cache

    @property
    def session_count(self) -> int:
        return len(self._sessions)

    def to_info(self) -> dict:
        return {
            "nodeId": self.node_id,
            "host": self.host,
            "port": self.port,
            "capabilities": self.capabilities,
            "connectedAt": self.connected_at.isoformat(),
            "sessionCount": self.session_count,
            "status": "connected",
        }

    # --- 명령 전송 ---

    def _next_request_id(self) -> str:
        self._request_counter += 1
        return f"req-{self._request_counter}-{int(time.time() * 1000)}"

    async def _send(self, data: dict) -> None:
        try:
            await self._ws.send_json(data)
        except Exception as e:
            # websockets 라이브러리가 close frame 이후 send를 시도하면
            # "Cannot call 'send' once a close message has been sent." RuntimeError를 발생시킨다.
            # 이를 WebSocketDisconnect로 정규화하여 호출자가 일관되게 처리할 수 있게 한다.
            raise WebSocketDisconnect(code=1011, reason=str(e)) from e

    async def _send_command(
        self, command: str, payload: dict, timeout: float = COMMAND_TIMEOUT
    ) -> dict:
        request_id = self._next_request_id()
        future: asyncio.Future[dict] = asyncio.get_running_loop().create_future()
        self._pending[request_id] = future

        message = {"type": command, "requestId": request_id, **payload}
        await self._send(message)

        try:
            return await asyncio.wait_for(future, timeout=timeout)
        except asyncio.TimeoutError:
            raise TimeoutError(
                f"Command {command} timed out after {timeout}s (request_id={request_id})"
            )
        finally:
            self._pending.pop(request_id, None)

    async def send_create_session(
        self,
        prompt: str,
        session_id: str | None = None,
        profile: str | None = None,
        allowed_tools: list[str] | None = None,
        disallowed_tools: list[str] | None = None,
        use_mcp: bool | None = None,
        node_id: str | None = None,
        folder_id: str | None = None,
    ) -> dict:
        payload: dict[str, Any] = {"prompt": prompt}
        if session_id:
            payload["agentSessionId"] = session_id
        if profile:
            payload["profile"] = profile
        if allowed_tools is not None:
            payload["allowedTools"] = allowed_tools
        if disallowed_tools is not None:
            payload["disallowedTools"] = disallowed_tools
        if use_mcp is not None:
            payload["useMcp"] = use_mcp
        if node_id is not None:
            payload["nodeId"] = node_id
        if folder_id is not None:
            payload["folderId"] = folder_id
        return await self._send_command(CMD_CREATE_SESSION, payload)

    async def send_intervene(
        self, session_id: str, text: str, user: str = ""
    ) -> dict:
        return await self._send_command(
            CMD_INTERVENE,
            {"agentSessionId": session_id, "text": text, "user": user},
        )

    async def send_respond(
        self, session_id: str, request_id: str, answers: dict
    ) -> dict:
        return await self._send_command(
            CMD_RESPOND,
            {
                "agentSessionId": session_id,
                "requestId": request_id,
                "answers": answers,
            },
        )

    async def send_subscribe_events(
        self, session_id: str, callback: Callable
    ) -> str:
        subscribe_id = str(uuid.uuid4())

        if session_id not in self._subscribe_listeners:
            self._subscribe_listeners[session_id] = {}
        self._subscribe_listeners[session_id][subscribe_id] = callback

        await self._send({
            "type": CMD_SUBSCRIBE_EVENTS,
            "agentSessionId": session_id,
            "subscribeId": subscribe_id,
        })
        return subscribe_id

    def unsubscribe_events(self, session_id: str, subscribe_id: str) -> None:
        listeners = self._subscribe_listeners.get(session_id)
        if listeners:
            listeners.pop(subscribe_id, None)
            if not listeners:
                del self._subscribe_listeners[session_id]

    # --- 수신 메시지 처리 ---

    async def handle_message(self, data: dict) -> None:
        msg_type = data.get("type")
        request_id = data.get("requestId")

        # pending request에 대한 응답
        if request_id and request_id in self._pending:
            future = self._pending.pop(request_id)
            if not future.done():
                if msg_type == EVT_ERROR:
                    future.set_exception(
                        RuntimeError(data.get("message", "Unknown error"))
                    )
                else:
                    future.set_result(data)
            return

        # 이벤트 디스패치
        if msg_type == EVT_SESSION_CREATED:
            await self._on_session_created(data)
        elif msg_type == EVT_EVENT:
            await self._on_event(data)
        elif msg_type == EVT_SESSIONS_UPDATE:
            await self._on_sessions_update(data)
        elif msg_type == EVT_SESSION_UPDATED:
            await self._on_session_updated(data)
        elif msg_type == EVT_SESSION_DELETED:
            await self._on_session_deleted(data)
        elif msg_type == EVT_HEALTH_STATUS:
            await self._on_health_status(data)
        elif msg_type == EVT_ERROR:
            logger.warning(
                "Error from node %s: %s", self.node_id, data.get("message")
            )
        else:
            logger.debug(
                "Unknown message type from node %s: %s", self.node_id, msg_type
            )

    async def _on_session_created(self, data: dict) -> None:
        session_id = data.get("agentSessionId")
        if session_id:
            self._sessions[session_id] = {
                "agentSessionId": session_id,
                "status": data.get("status", "running"),
                "nodeId": self.node_id,
            }
            if self.on_session_change:
                await self.on_session_change(
                    self.node_id, "session_created", data
                )

    async def _on_event(self, data: dict) -> None:
        session_id = data.get("agentSessionId") or data.get("sessionId")
        subscribe_id = data.get("subscribeId")

        if session_id and session_id in self._subscribe_listeners:
            listeners = self._subscribe_listeners[session_id]
            if subscribe_id and subscribe_id in listeners:
                await listeners[subscribe_id](data)
            else:
                # broadcast to all listeners for this session
                for cb in list(listeners.values()):
                    await cb(data)

    async def _on_sessions_update(self, data: dict) -> None:
        sessions = data.get("sessions", [])
        self._sessions.clear()
        for s in sessions:
            sid = s.get("agentSessionId") or s.get("session_id")
            if sid:
                self._sessions[sid] = s
        if self.on_session_change:
            await self.on_session_change(
                self.node_id, "sessions_update", data
            )

    async def _on_session_updated(self, data: dict) -> None:
        session_id = data.get("agentSessionId") or data.get("session_id")
        if session_id and session_id in self._sessions:
            self._sessions[session_id].update(data)
        if self.on_session_change:
            await self.on_session_change(
                self.node_id, "session_updated", data
            )

    async def _on_session_deleted(self, data: dict) -> None:
        session_id = data.get("agentSessionId") or data.get("session_id")
        if session_id:
            self._sessions.pop(session_id, None)
        if self.on_session_change:
            await self.on_session_change(
                self.node_id, "session_deleted", data
            )

    async def _on_health_status(self, data: dict) -> None:
        logger.debug("Health status from node %s: %s", self.node_id, data)

    # --- 연결 종료 ---

    async def close(self) -> None:
        # cancel all pending futures
        for future in self._pending.values():
            if not future.done():
                future.cancel()
        self._pending.clear()
        self._subscribe_listeners.clear()

        try:
            await self._ws.close()
        except Exception:
            pass

        if self.on_close:
            cb = self.on_close
            self.on_close = None  # 이중 호출 방지 — ws_handler finally와 register_node 중복 호출
            await cb(self)
